Fix `colcon test`

Martin-Idel opened the pull request:
Rename certain files to hpp because uncrustify wants to completely change indentation and cpplint is unhappy about this.

This should hopefully fix all errors in CI.

If you want, I can also rename all headers to hpp (also in image_transport) and enable cppcheck. I didn't do it immediately to make the changes as small as possible.

Martin-Idel committed:
Fix uncrustify

Rename certain files to hpp because uncrustify wants
to completely change indentation and cpplint is unhappy
about this

mjcarroll commented:
With renaming those headers, this will definitely have to land in `foxy`.

Is there any way to do a deprecation cycle on the current header locations?

Martin-Idel committed:
Add deprecation notes

Martin-Idel commented:
I would have assumed this landed in foxy - so for me that'd be okay ;-) .

However, I can understand if you also want to make it available in eloquent or earlier, so here is what I see we could do:
- Either disable uncrustify for those files and keep them as `.h` files. I don't particularly like this, but it's possible. 
- Or rename the headers to `.hpp` and add a new file with `.h` which just includes the `.hpp` file and adds a deprecation warning. I added a commit to show how I'd do this. If you want, I can quickly do this for all headers. The build still compiles and is error free. Any build I can see which uses the `.h`file will then still compile but issue a "note" on the command line 
I'm not exactly sure that this is completely portable though but at least there is no warning from `-Wcpp`.

CI failure is unrelated, I can't reproduce this and I also didn't really change `image_transport` in this last commit and the removal of the filesystem_helper fixes this again.

Martin-Idel committed:
Try to remove filesystem helper

Martin-Idel commented:
I also saw that the windows build fails due to problems with the filesystem helper.

My solution would be to delete it and use `rcpputils/filesystem_helper`. From a library perspective, this should still make it possible to backport, if `rcpputils` still exists.

mjcarroll commented:
> My solution would be to delete it and use rcpputils/filesystem_helper.

I agree with that decision, and that's available in previous versions as well.

mjcarroll commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=9890)](http://ci.ros2.org/job/ci_linux/9890/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5507)](http://ci.ros2.org/job/ci_linux-aarch64/5507/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8047)](http://ci.ros2.org/job/ci_osx/8047/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9746)](http://ci.ros2.org/job/ci_windows/9746/)

Martin-Idel committed:
Change method return type to size_t

Fixes MSVC warning about conversion from size_t

Martin-Idel committed:
Fix MSVC warning for getenv

Do it similar to ament_index_cpp

Martin-Idel commented:
Of course there had to be some MSVC warnings...
Hopefully fixed them all.

Regarding the header switch/deprecation, do you want me to also implement it for the `image_transport` subproject? If yes, would you like a followup-PR or should I do it here?

mjcarroll commented:
> Of course there had to be some MSVC warnings...

Always ;)

> If yes, would you like a followup-PR or should I do it here?

Yes, and here is fine, since we need at least one more CI run, and currently waiting to bring buildfarm back until the outage is over, so we have a bit of time.

Martin-Idel committed:
Rename .h files to .hpp + formatting fixes

- fix include guards to cpplint style
- fix order of includes to match ros guide

Martin-Idel committed:
Add back deprecated .h headers

Martin-Idel commented:
So I added the `.h` deprecated headers to `image_transport` - except for the tutorials and tests, where I only renamed the headers to `.hpp`. 

Since I had to adapt the header guards anyways, I also updated them to the style preferred by cpplint. Some includes got moved around by my IDE in the process of renaming files.

In order to make everything still somewhat reviewable I left the commits separated (the PR overview is completely messed up). Since I had to rebase on master, the two new commits are the last two of the bunch.

To test it (at least on Linux), I build `rviz_default_plugins` with this PR https://github.com/ros2/rviz/pull/523 still containing the `.h` headers.
`colcon build` runs without errors but as expected I get stderr output:
```
[Processing: rviz_default_plugins]                                        
--- stderr: rviz_default_plugins                                              
In file included from /home/martin/gitRepos/ros/rviz_ws/src/rviz_default_plugins/include/rviz_default_plugins/displays/image/image_transport_display.hpp:40:0,
                 from /home/martin/gitRepos/ros/rviz_ws/src/rviz_default_plugins/include/rviz_default_plugins/displays/camera/camera_display.hpp:51,
                 from /home/martin/gitRepos/ros/rviz_ws/src/rviz_default_plugins/src/rviz_default_plugins/displays/camera/camera_display.cpp:32:
/home/martin/gitRepos/ros/image_transport/install/include/image_transport/image_transport.h:41:89: note: #pragma message: Warning: This header is deprecated. Use 'image_transport.hpp' instead
 #pragma message ("Warning: This header is deprecated. Use 'image_transport.hpp' instead")
```
So it's hard to ignore but not a compiler warning (because I didn't find a portable way that isn't compiler specific)

mjcarroll commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10021)](http://ci.ros2.org/job/ci_linux/10021/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5620)](http://ci.ros2.org/job/ci_linux-aarch64/5620/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8155)](http://ci.ros2.org/job/ci_osx/8155/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9927)](http://ci.ros2.org/job/ci_windows/9927/)

Martin-Idel committed:
Change all getNumSubscriber/Publisher signatures to size_t

Martin-Idel committed:
Fix leaky windows workaround for getenv

Martin-Idel commented:
Argh... I did another pass at fixing MSVC warnings - I grepped for all `uint32_t` and found a number of further cases which originate from the node signature using a `size_t` (all transferred to `size_t`).

Regarding the test failures, this might be related to my replacement of `getenv`, at least the old implementation is wrong: It leaked a `char *` pointer and there was also UB if the environment variable was empty.

Trouble is, I don't currently have a local Windows installation to test this, which would make turnaround time shorter.

EDIT: But it turns out I still have CI access, so let's try Windows again (hopefully didn't mess up the parameters):
[![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9934)](http://ci.ros2.org/job/ci_windows/9934/)

Martin-Idel commented:
Actually, the test failures are weird - they are only concerned with `image_transport`, where I didn't change the logic at all and they seem to also change from run to run (filesystem helper and getenv changes are for one of the other packages).

I guess I'll have to actually debug this on Windows, which means I have to get my build working again, which will probably take a couple of days.

Martin-Idel commented:
So the test failures seem to already be present on master (at least on my machine). I'm trying to get a debug build working on Windows, but that's complicated...

The errors are weird - it seems to me as if some environment variables were wrong, but they aren't. Do you happen to have a good idea @mjcarroll. Maybe that could accelerate things...

mjcarroll commented:
@Martin-Idel thanks for diving in, I'll try to take a look, but may be slightly delayed due to getting my Windows VM up-to-date.  

The failing to find symbol warnings are quite suspicious, though.  It may be that new symbols have been added without setting the visibility correctly?

Martin-Idel commented:
@mjcarroll No, this will have nothing to do with visibility - because the symbols we are talking about are `sensor_msgs` symbols and if those are missing, the problem will be much more widespread.

I tried to do some more digging - but I can't get a debug build to work nor attach symbol files to everything I need and therefore I can't evaluate any further. 

Here is what I can see:
The following seems to work perfectly (this is also basically the test_subscriber - as test fixture take the one e.g. from the MessagePassingTest):

```
TEST_F(SomeTest, test1)
{
  std::function<void(const sensor_msgs::msg::Image::ConstSharedPtr & msg)> imageCallback =
    [](const auto & msg) {(void)msg;};

  auto sub = image_transport::create_subscription(node_.get(), "camera/image", imageCallback, "raw");
  rclcpp::executors::SingleThreadedExecutor executor;

  test_rclcpp::wait_for_subscriber(node_, sub.getTopic());
  executor.spin_node_some(node_);
}
```
You can also add new subscribers, wait some more, spin some more, nothing seems to be a problem.
Similarly, the following works perfectly on my machine:
```
TEST_F(SomeTest, test2)
{
  std::function<void(const sensor_msgs::msg::Image::ConstSharedPtr & msg)> imageCallback =
    [](const auto & msg) {(void)msg;};
  auto pub = image_transport::create_publisher(node_.get(), "camera/image");
  
  pub.publish(generate_random_image());
  executor.spin_node_some(node_);
}
```
You can also add some publishers, wait, publish again, spin again.
However, if you combine this, it doesn't work:
```
TEST_F(SomeTest, test2)
{
  std::function<void(const sensor_msgs::msg::Image::ConstSharedPtr & msg)> imageCallback =
    [](const auto & msg) {(void)msg;};
  auto pub = image_transport::create_publisher(node_.get(), "camera/image");
  auto sub = image_transport::create_subscription(node_.get(), "camera/image", imageCallback, "raw");
  rclcpp::executors::SingleThreadedExecutor executor;

  test_rclcpp::wait_for_subscriber(node_, sub.getTopic());
  executor.spin_node_some(node_);
}
```
This breaks deterministically on my machine. This tells me that this is somehow a tricky problem with the image publishers/subscribers. 

Since it's limited to Windows the first idea would be undefined behaviour. I ran ubsan on Linux and it does find something in the SubscriberOptions within `rclcpp` but I have no idea whether this is related. I also ran asan which found a lot of leakage, all within pluginlib (registration of plugins, I think this is known) as well as within rmw_fastrtps. So nothing helping here. Very unlucky.

Martin-Idel commented:
@mjcarroll I got my debug build to work, but I still can't pinpoint the problem and now I'm out of ideas. Maybe the below analysis makes sense to you?

So what the stacktrace is telling me that indeed, the subscription cannot be initialized. In fact, the problem seems to be (as the error messages says) when trying to get the typesupport (and then a second error gets thrown on deletion of the subscription we are about to create after the error is thrown) - but I don't get why.

I tried to see differences between the subscriber test (which passes and correctly constructs the subscription) and the message passing test, which doesn't. 

Here is what I see with the message passing test:
- I get to https://github.com/ros2/rmw_fastrtps/blob/289c3462bd568a03318140b3e05ca4a1ae9cd6eb/rmw_fastrtps_cpp/src/subscription.cpp#L82 where we try to obtain typesupport
- We are asking for typsupport `0x00007ffc4aa092a0 {sensor_msgs__rosidl_typesupport_cpp.dll!rosidl_message_type_support_t sensor_msgs::msg::rosidl_typesupport_cpp::Image_message_type_support_handle} {...}` but we don't get it. 
- I can't really debug into that without a lot of effort (only get disassembly) but from what I gather, the `get_message_typesupport_handle` tries to search for the symbols in a bunch of libraries it stores in some map. Those entries look good (containing stuff concerning `sensor_msgs`).

Here is what I see with the subscription test:
- I get to https://github.com/ros2/rmw_fastrtps/blob/289c3462bd568a03318140b3e05ca4a1ae9cd6eb/rmw_fastrtps_cpp/src/subscription.cpp#L82 where we try to obtain typesupport
- We are asking for typsupport `0x00007ffc4aa092a0 {sensor_msgs__rosidl_typesupport_cpp.dll!rosidl_message_type_support_t sensor_msgs::msg::rosidl_typesupport_cpp::Image_message_type_support_handle} {...}` but we don't get it. 
- For some reason I can debug into that a little bit further: I see the `type_support_dispatch.hpp` where the `type_support_mmap_t` map name `map` seems to contain the exact same entries as the assembly parts the debugger identifies as `map` in the message passing test. But I can't really debug here any further without running into assembly, too.

So the difference is that in one instance we get the correct typesupport handle deep down in fastrtps and in the other we don't. Superficially though, everything seems to be the same. But that's where I'm stuck. Maybe there is indeed some undefined behaviour somewhere? Maybe the underlying objects get destroyed although they shouldn't (a superficial glance doesn't tell me anything).

Finally, the stack traces in that particular moment are identical (above the test stack trace). It's also not how the tests are set up (includes, environment, etc.). If I just change the content of the `TEST_F` macro with the corresponding other content, the same behaviour occurs.

So I'm back with thinking it might be some local that gets destroyed although it shouldn't  or some undefined behaviour somewhere.

(Here is the subscriber stack trace, everything above the last entry reading `image_transport-subscriber.exe` is the same in both tests at the moment where getting the typesupport fails:

>	rmw_fastrtps_cpp.dll!rmw_fastrtps_cpp::create_subscription(const CustomParticipantInfo * participant_info, const rosidl_message_type_support_t * type_supports, const char * topic_name, const rmw_qos_profile_t * qos_policies, const rmw_subscription_options_t * subscription_options, bool keyed, bool create_subscription_listener) Line 82	C++
 	rmw_fastrtps_cpp.dll!rmw_create_subscription(const rmw_node_t * node, const rosidl_message_type_support_t * type_supports, const char * topic_name, const rmw_qos_profile_t * qos_policies, const rmw_subscription_options_t * subscription_options) Line 76	C++
 	rmw_implementation.dll!rmw_create_subscription(const rmw_node_t * v5, const rosidl_message_type_support_t * v4, const char * v3, const rmw_qos_profile_t * v2, const rmw_subscription_options_t * v1) Line 325	C++
 	rcl.dll!rcl_subscription_init(rcl_subscription_t * subscription, const rcl_node_t * node, const rosidl_message_type_support_t * type_support, const char * topic_name, const rcl_subscription_options_t * options) Line 161	C
 	rclcpp.dll!rclcpp::SubscriptionBase::SubscriptionBase(rclcpp::node_interfaces::NodeBaseInterface * node_base, const rosidl_message_type_support_t & type_support_handle, const std::string & topic_name, const rcl_subscription_options_t & subscription_options, bool is_serialized) Line 62	C++
 	image_transport_plugins.dll!rclcpp::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>(rclcpp::node_interfaces::NodeBaseInterface * node_base, const rosidl_message_type_support_t & type_support_handle, const std::string & topic_name, const rclcpp::QoS & qos, rclcpp::AnySubscriptionCallback<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>> callback, const rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> & options, std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> message_memory_strategy) Line 101	C++
 	image_transport_plugins.dll!std::_Construct_in_place<rclcpp::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>,rclcpp::node_interfaces::NodeBaseInterface * &,rosidl_message_type_support_t const &,std::string const &,rclcpp::QoS const &,rclcpp::AnySubscriptionCallback<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>> const &,rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> const &,std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> const &>(rclcpp::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> & _Obj, rclcpp::node_interfaces::NodeBaseInterface * & <_Args_0>, const rosidl_message_type_support_t & <_Args_1>, const std::string & <_Args_2>, const rclcpp::QoS & <_Args_3>, const rclcpp::AnySubscriptionCallback<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>> & <_Args_4>, const rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> & <_Args_5>, const std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> & <_Args_6>) Line 205	C++
 	image_transport_plugins.dll!std::_Ref_count_obj2<rclcpp::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>>::_Ref_count_obj2<rclcpp::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>><rclcpp::node_interfaces::NodeBaseInterface * &,rosidl_message_type_support_t const &,std::string const &,rclcpp::QoS const &,rclcpp::AnySubscriptionCallback<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>> const &,rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> const &,std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> const &>(rclcpp::node_interfaces::NodeBaseInterface * & <_Args_0>, const rosidl_message_type_support_t & <_Args_1>, const std::string & <_Args_2>, const rclcpp::QoS & <_Args_3>, const rclcpp::AnySubscriptionCallback<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>> & <_Args_4>, const rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> & <_Args_5>, const std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> & <_Args_6>) Line 1486	C++
 	image_transport_plugins.dll!std::make_shared<rclcpp::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>,rclcpp::node_interfaces::NodeBaseInterface * &,rosidl_message_type_support_t const &,std::string const &,rclcpp::QoS const &,rclcpp::AnySubscriptionCallback<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>> const &,rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> const &,std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> const &>(rclcpp::node_interfaces::NodeBaseInterface * & <_Args_0>, const rosidl_message_type_support_t & <_Args_1>, const std::string & <_Args_2>, const rclcpp::QoS & <_Args_3>, const rclcpp::AnySubscriptionCallback<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>> & <_Args_4>, const rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> & <_Args_5>, const std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> & <_Args_6>) Line 1584	C++
 	image_transport_plugins.dll!rclcpp::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>::make_shared<rclcpp::node_interfaces::NodeBaseInterface * &,rosidl_message_type_support_t const &,std::string const &,rclcpp::QoS const &,rclcpp::AnySubscriptionCallback<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>> const &,rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> const &,std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> const &>(rclcpp::node_interfaces::NodeBaseInterface * & <args_0>, const rosidl_message_type_support_t & <args_1>, const std::string & <args_2>, const rclcpp::QoS & <args_3>, const rclcpp::AnySubscriptionCallback<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>> & <args_4>, const rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> & <args_5>, const std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> & <args_6>) Line 78	C++
 	image_transport_plugins.dll!rclcpp::create_subscription_factory::__l2::<lambda>(rclcpp::node_interfaces::NodeBaseInterface * node_base, const std::string & topic_name, const rclcpp::QoS & qos) Line 100	C++
 	image_transport_plugins.dll!std::_Invoker_functor::_Call<std::shared_ptr<rclcpp::SubscriptionBase> <lambda>(rclcpp::node_interfaces::NodeBaseInterface *, const std::string &, const rclcpp::QoS &) &,rclcpp::node_interfaces::NodeBaseInterface *,std::string const &,rclcpp::QoS const &>(rclcpp::create_subscription_factory::__l2::std::shared_ptr<rclcpp::SubscriptionBase> <lambda>(rclcpp::node_interfaces::NodeBaseInterface *, const std::string &, const rclcpp::QoS &) & _Obj, rclcpp::node_interfaces::NodeBaseInterface * && <_Args_0>, const std::string & <_Args_1>, const rclcpp::QoS & <_Args_2>) Line 1610	C++
 	image_transport_plugins.dll!std::invoke<std::shared_ptr<rclcpp::SubscriptionBase> <lambda>(rclcpp::node_interfaces::NodeBaseInterface *, const std::string &, const rclcpp::QoS &) &,rclcpp::node_interfaces::NodeBaseInterface *,std::string const &,rclcpp::QoS const &>(rclcpp::create_subscription_factory::__l2::std::shared_ptr<rclcpp::SubscriptionBase> <lambda>(rclcpp::node_interfaces::NodeBaseInterface *, const std::string &, const rclcpp::QoS &) & _Obj, rclcpp::node_interfaces::NodeBaseInterface * && <_Args_0>, const std::string & <_Args_1>, const rclcpp::QoS & <_Args_2>) Line 1610	C++
 	image_transport_plugins.dll!std::_Invoker_ret<std::shared_ptr<rclcpp::SubscriptionBase>,0>::_Call<std::shared_ptr<rclcpp::SubscriptionBase> <lambda>(rclcpp::node_interfaces::NodeBaseInterface *, const std::string &, const rclcpp::QoS &) &,rclcpp::node_interfaces::NodeBaseInterface *,std::string const &,rclcpp::QoS const &>(rclcpp::create_subscription_factory::__l2::std::shared_ptr<rclcpp::SubscriptionBase> <lambda>(rclcpp::node_interfaces::NodeBaseInterface *, const std::string &, const rclcpp::QoS &) & <_Vals_0>, rclcpp::node_interfaces::NodeBaseInterface * && <_Vals_1>, const std::string & <_Vals_2>, const rclcpp::QoS & <_Vals_3>) Line 1636	C++
 	image_transport_plugins.dll!std::_Func_impl_no_alloc<std::shared_ptr<rclcpp::SubscriptionBase> <lambda>(rclcpp::node_interfaces::NodeBaseInterface *, const std::string &, const rclcpp::QoS &),std::shared_ptr<rclcpp::SubscriptionBase>,rclcpp::node_interfaces::NodeBaseInterface *,std::string const &,rclcpp::QoS const &>::_Do_call(rclcpp::node_interfaces::NodeBaseInterface * && <_Args_0>, const std::string & <_Args_1>, const rclcpp::QoS & <_Args_2>) Line 926	C++
 	rclcpp.dll!std::_Func_class<std::shared_ptr<rclcpp::SubscriptionBase>,rclcpp::node_interfaces::NodeBaseInterface *,std::string const &,rclcpp::QoS const &>::operator()(rclcpp::node_interfaces::NodeBaseInterface * <_Args_0>, const std::string & <_Args_1>, const rclcpp::QoS & <_Args_2>) Line 976	C++
 	rclcpp.dll!rclcpp::node_interfaces::NodeTopics::create_subscription(const std::string & topic_name, const rclcpp::SubscriptionFactory & subscription_factory, const rclcpp::QoS & qos) Line 78	C++
 	image_transport_plugins.dll!rclcpp::create_subscription<sensor_msgs::msg::Image_<std::allocator<void>>,void <lambda>(const std::shared_ptr<sensor_msgs::msg::Image_<std::allocator<void>> const>),std::allocator<void>,sensor_msgs::msg::Image_<std::allocator<void>>,rclcpp::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>,rclcpp::Node &>(rclcpp::Node & node, const std::string & topic_name, const rclcpp::QoS & qos, image_transport::SimpleSubscriberPlugin<sensor_msgs::msg::Image_<std::allocator<void>>>::subscribeImpl::__l2::void <lambda>(const std::shared_ptr<sensor_msgs::msg::Image_<std::allocator<void>> const>) && callback, const rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> & options, std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> msg_mem_strat) Line 73	C++
 	image_transport_plugins.dll!rclcpp::Node::create_subscription<sensor_msgs::msg::Image_<std::allocator<void>>,void <lambda>(const std::shared_ptr<sensor_msgs::msg::Image_<std::allocator<void>> const>),std::allocator<void>,sensor_msgs::msg::Image_<std::allocator<void>>,rclcpp::Subscription<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>,rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>>(const std::string & topic_name, const rclcpp::QoS & qos, image_transport::SimpleSubscriberPlugin<sensor_msgs::msg::Image_<std::allocator<void>>>::subscribeImpl::__l2::void <lambda>(const std::shared_ptr<sensor_msgs::msg::Image_<std::allocator<void>> const>) && callback, const rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> & options, std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<sensor_msgs::msg::Image_<std::allocator<void>>,std::allocator<void>>> msg_mem_strat) Line 94	C++
 	image_transport_plugins.dll!image_transport::SimpleSubscriberPlugin<sensor_msgs::msg::Image_<std::allocator<void>>>::subscribeImpl(rclcpp::Node * node, const std::string & base_topic, const std::function<void __cdecl(std::shared_ptr<sensor_msgs::msg::Image_<std::allocator<void>> const> const &)> & callback, rmw_qos_profile_t custom_qos) Line 125	C++
 	image_transport.dll!image_transport::SubscriberPlugin::subscribe(rclcpp::Node * node, const std::string & base_topic, const std::function<void __cdecl(std::shared_ptr<sensor_msgs::msg::Image_<std::allocator<void>> const> const &)> & callback, rmw_qos_profile_t custom_qos) Line 76	C++
 	image_transport.dll!image_transport::Subscriber::Subscriber(rclcpp::Node * node, const std::string & base_topic, const std::function<void __cdecl(std::shared_ptr<sensor_msgs::msg::Image_<std::allocator<void>> const> const &)> & callback, std::shared_ptr<pluginlib::ClassLoader<image_transport::SubscriberPlugin>> loader, const std::string & transport, rmw_qos_profile_t custom_qos) Line 128	C++
 	image_transport.dll!image_transport::create_subscription(rclcpp::Node * node, const std::string & base_topic, const std::function<void __cdecl(std::shared_ptr<sensor_msgs::msg::Image_<std::allocator<void>> const> const &)> & callback, const std::string & transport, rmw_qos_profile_t custom_qos) Line 78	C++
 	image_transport-subscriber.exe!TestPublisher_construction_and_destruction_Test::TestBody() Line 27	C++
 	image_transport-subscriber.exe!testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2432	C++
 	image_transport-subscriber.exe!testing::internal::HandleExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2483	C++
 	image_transport-subscriber.exe!testing::Test::Run() Line 2529	C++
 	image_transport-subscriber.exe!testing::TestInfo::Run() Line 2707	C++
 	image_transport-subscriber.exe!testing::TestCase::Run() Line 2826	C++
 	image_transport-subscriber.exe!testing::internal::UnitTestImpl::RunAllTests() Line 5217	C++
 	image_transport-subscriber.exe!testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>(testing::internal::UnitTestImpl * object, bool(testing::internal::UnitTestImpl::*)() method, const char * location) Line 2432	C++
 	image_transport-subscriber.exe!testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>(testing::internal::UnitTestImpl * object, bool(testing::internal::UnitTestImpl::*)() method, const char * location) Line 2483	C++
 	image_transport-subscriber.exe!testing::UnitTest::Run() Line 4824	C++
 	image_transport-subscriber.exe!RUN_ALL_TESTS() Line 2371	C++
 	image_transport-subscriber.exe!main(int argc, char * * argv) Line 38	C++
 	image_transport-subscriber.exe!invoke_main() Line 79	C++
 	image_transport-subscriber.exe!__scrt_common_main_seh() Line 288	C++
 	image_transport-subscriber.exe!__scrt_common_main() Line 331	C++
 	image_transport-subscriber.exe!mainCRTStartup() Line 17	C++
 	kernel32.dll!00007ffc64d27bd4()	Unknown
 	ntdll.dll!00007ffc655ace51()	Unknown

clalancette commented on camera_info_manager/src/camera_info_manager.cpp:
```diff
@@ -313,6 +316,23 @@ bool CameraInfoManager::loadCameraInfo(const std::string & url)
   return loadCalibration(url, cname);
 }
 
+std::string get_env(const char * env_var)
+{
+#ifndef _WIN32
+  auto env_variable = getenv(env_var);
+#else
+  char * output_env_var = nullptr;
+  size_t env_var_size;
+  _dupenv_s(&output_env_var, &env_var_size, env_var);
+
+  std::string env_variable = "";
+  if (output_env_var) {
+    env_variable = std::string(output_env_var);
+    free(output_env_var);
+  }
+#endif
+  return env_variable;
+}
```
Can I suggest using `rcutils_get_env` here instead?  Especially since this is new code.

mjcarroll replied:
I'm going to use the `rcpputils` variant, since we already have that dependency.

clalancette commented on image_transport/include/image_transport/single_subscriber_publisher.hpp:
```diff
@@ -0,0 +1,98 @@
+
+#include <sensor_msgs/msg/image.hpp>
+
+#include <string>
+#include <functional>
```
Nit: alphabetize.

clalancette commented on image_transport/include/image_transport/subscriber.hpp:
```diff
@@ -0,0 +1,123 @@
+
+#ifndef IMAGE_TRANSPORT__SUBSCRIBER_HPP_
+#define IMAGE_TRANSPORT__SUBSCRIBER_HPP_
+
```
Nit:

```
#include <functional>
#include <memory>
#include <string>
```

clalancette commented on image_transport/include/image_transport/subscriber_filter.hpp:
```diff
@@ -0,0 +1,175 @@
+
+#ifndef IMAGE_TRANSPORT__SUBSCRIBER_FILTER_HPP_
+#define IMAGE_TRANSPORT__SUBSCRIBER_FILTER_HPP_
+
```
Nit:

```
#include <string>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
```

clalancette reviewed (commented):
I left a few nits to fix, but nothing really major stood out to me.

clalancette committed:
Fixes to ament dependencies.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

Commit:
Use rcpputils::get_env_var rather than custom

Signed-off-by: Michael Carroll <michael@openrobotics.org>

Commit:
Address reviewer feedback

Signed-off-by: Michael Carroll <michael@openrobotics.org>

Commit:
Re-add rcpputils for get_env_var

Signed-off-by: Michael Carroll <michael@openrobotics.org>

mjcarroll commented:
With latest changes, tests pass for @clalancette locally (I'm having some timeouts).

CI run:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10912)](http://ci.ros2.org/job/ci_linux/10912/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6264)](http://ci.ros2.org/job/ci_linux-aarch64/6264/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8887)](http://ci.ros2.org/job/ci_osx/8887/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10809)](http://ci.ros2.org/job/ci_windows/10809/)

Commit:
Change to new style remapping

Signed-off-by: Michael Carroll <michael@openrobotics.org>

Commit:
Move dllimport/export locations

Signed-off-by: Michael Carroll <michael@openrobotics.org>

mjcarroll commented:
After addressing `camera_info_manager` dllimport:

[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10811)](https://ci.ros2.org/job/ci_windows/10811/)

Martin-Idel commented:
@mjcarroll @clalancette Thanks for finding and fixing my last issues! Much appreciated.

mjcarroll commented:
Thanks for iterating on it! Windows can be a bit frustrating sometimes ;)

dirk-thomas commented:
This change introduced pragma warnings when building `rqt_image_view` which have been addressed in ros-visualization/rqt_image_view#34.
